services/src/main/java/org/keycloak/services/DefaultKeycloakContext.java

```diff
@@ -113,6 +116,11 @@ public void setRealm(RealmModel realm) {
 
     @Override
     public ClientModel getClient() {
+        if (client == null) {
```

Comment:
If it makes sense, I can add a check to the client set by `setClient` is always the same from the authentication session, if the latter is set.